samples/nrf9160/serial_lte_modem/src/slm_at_ftp.c

```diff
@@ -0,0 +1,520 @@
+{
+	rsp_send((u8_t *)msg, len);
+}
+
+void ftp_data_callback(const u8_t *msg, u16_t len)
+{
+	if (slm_util_hex_check((u8_t *)msg, len)) {
+		int ret;
+		int size = len * 2;
+
+		ret = slm_util_htoa(msg, len, rsp_buf, size);
+		if (ret > 0) {
+			rsp_send(rsp_buf, ret);
+		} else {
+			LOG_WRN("hex convert error: %d", ret);
+		}
+	} else {
+		rsp_send((u8_t *)msg, len);
+	}
+}
+
+/* AT#XFTP="open",<username>,<password>,<hostname>[,<port>[,<sec_tag>]] */
+/* similar to ftp://<user>:<password>@<host>:<port> */
+static int do_ftp_open(void)
+{
+	int ret = 0;
+	char username[FTP_MAX_USERNAME];
+	int sz_username = FTP_MAX_USERNAME;
+	char password[FTP_MAX_PASSWORD];
+	int sz_password = FTP_MAX_PASSWORD;
+	char hostname[FTP_MAX_HOSTNAME];
+	int sz_hostname = FTP_MAX_HOSTNAME;
+	u16_t port = CONFIG_FTP_SERVER_PORT;
+	sec_tag_t sec_tag = INVALID_SEC_TAG;
+	int param_count;
+
+	/* Parse AT command */
+	param_count = at_params_valid_count_get(&at_param_list);
+	if (param_count < 5) {
+		return -EINVAL;
+	}
+	memset(username, 0x00, sz_username); /* Important for optional params*/
+	ret = at_params_string_get(&at_param_list, 2, username, &sz_username);
+	if (ret || strlen(username) == 0) {
+		memcpy(username, CONFIG_FTP_USER_ANONYMOUS,
+			sizeof(CONFIG_FTP_USER_ANONYMOUS) - 1);
+		memcpy(password, CONFIG_FTP_PASSWORD_ANONYMOUS,
+			sizeof(CONFIG_FTP_PASSWORD_ANONYMOUS) - 1);
+	} else {
+		username[sz_username] = '\0';
+		at_params_string_get(&at_param_list, 3, password, &sz_password);
+		password[sz_password] = '\0';
+	}
+	ret = at_params_string_get(&at_param_list, 4, hostname, &sz_hostname);
+	if (ret) {
+		return ret;
+	}
+	hostname[sz_hostname] = '\0';
+	if (param_count > 5) {
+		ret = at_params_short_get(&at_param_list, 5, &port);
+		if (ret) {
+			return ret;
+		}
+	}
+	if (param_count > 6) {
+		ret = at_params_int_get(&at_param_list, 6, &sec_tag);
+		if (ret) {
+			return ret;
+		}
+	}
+
+	/* FTP open */
+	ret = ftp_open(hostname, port, sec_tag);
+	if (ret != FTP_CODE_200) {
+		return -ENETUNREACH;
+	}
+
+	/* FTP login */
+	ret = ftp_login(username, password);
+	if (ret != FTP_CODE_230) {
+		return -EACCES;
+	}
+
+	return 0;
+}
+
+/* AT#XFTP="close" */
+static int do_ftp_close(void)
+{
+	int ret = ftp_close();
+
+	return (ret == FTP_CODE_221)?0 : -1;
+}
+
+/* AT#XFTP="status" */
+static int do_ftp_status(void)
+{
+	int ret = ftp_status();
+
+	return (ret == FTP_CODE_211)?0 : -1;
+}
+
+/* AT#XFTP="ascii" */
+static int do_ftp_ascii(void)
+{
+	int ret = ftp_type(FTP_TYPE_ASCII);
+
+	return (ret == FTP_CODE_200)?0 : -1;
+}
+
+/* AT#XFTP="binary" */
+static int do_ftp_binary(void)
+{
+	int ret = ftp_type(FTP_TYPE_BINARY);
+
+	return (ret == FTP_CODE_200)?0 : -1;
+}
+
+/* AT#XFTP="pwd" */
+static int do_ftp_pwd(void)
+{
+	int ret = ftp_pwd();
+
+	return (ret == FTP_CODE_257)?0 : -1;
+}
+
+/* AT#XFTP="ls"[,<options>[,<folder>]] */
+static int do_ftp_ls(void)
+{
+	int ret;
+	char options[32];
+	int sz_options = 32;
+	char target[128];
```

Comment:
Could we add some constants for the buffer sizes? Something like `FTP_MAX_PATH_LEN`, this could be reused in several functions instead of magic numbers.

Reply:
Fixed